### PR TITLE
Fixes #3012 Fixes #3119 Digital Object Identifiers to hyperlinks

### LINKED ISSUE #3012: Publication: Allow Digital Object Identifier (DOI) to render as a LINK instead of a plain text.
**Do not disclose security vulnerabilities here, please email us instead: 
az-digital-security@list.arizona.edu**

## Motivation
With this feature, the DOI within a Publication will take users directly to the information page for the article.

### Is your feature request related to a problem? Please describe.
Unlike with Biblio where DOIs are rendered as links (i.e., https://ccentermdev-cancer-center-quickstart-2.pantheonsite.io/biblio), in Publication DOIs are rendered as plain texts (i.e., https://cancercenter.arizona.edu/biblio).

### Describe the solution you'd like
Format the DOI field to be presented as a link after you create a Publication. 

**For example:**

- https://ccentermdev-cancer-center-quickstart-2.pantheonsite.io/biblio
- https://ccentermdev-cancer-center-quickstart-2.pantheonsite.io/biblio/abstract-5938-administering-colorectal-cancer-education-inflatable-colon-underrepresented-and

## Additional context
Screenshot 2023-12-13 at 10 11 24 AM

Screenshot 2023-12-13 at 10 10 41 AM
Screenshot 2023-12-13 at 10 09 16 AM

### LINKED ISSUE #3119: Make URLs in Publication Citations into hyperlinks
## Motivation
During the discussion of #3012 it was discussed that we would make URLS in publication citations into links, which is partially motivated by the fact that some citation styles render DOI identifiers as links. However, this also would be a feature enhancement outside of DOI identifiers.

## Proposed Resolution
Make URLs into hyperlinks in the new proposed `az_citation` text format.

### Describe alternatives you've considered
Not having URLs become hyperlinks, which would have some implications for styles that choose to render DOI identifiers as URLs.

### PR DESCRIPTION
**Pull request by tadean**
## Description
This PR causes publication DOI identifiers in references to become hyperlinks. The approach taken is one of text format recognition rather than themeing - it attempts to identify DOIs in the reference output rather than making changes to the styles. The goal of this is to avoid altering the original styles unless we have to - otherwise this functionality would require customizing all citation styles that are part of Quickstart. The approach to take on this should probably be discussed to see if this seems reasonable.

This PR operates by providing a new text format (`az_citation`) that makes use of a custom Filter plugin provided by this PR, which recognizes DOIs by regular expression. At present, the text format contains filters that turn DOIs into links, and turns URLs into hyperlinks. Other publication-specific format tweaking could potentially be added to this text format in the future.

## Related issues
Closes #3012 
Closes #3119 

## How to test
- Enable `az_publication`
- Create publications with valid DOIs
- Verify they are turned into hyperlinks

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [x] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by sanjanans — APPROVED** [review]

File: modules/custom/az_publication/src/Plugin/Filter/AZDOIFilter.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  public function process($text, $langcode) {
+    // Based upon _filter_url().
+    // Store the current text in case any of the preg_* functions fail.
+    $saved_text = $text;
+
+    // Tags to skip and not recurse into.
+    $ignore_tags = 'a|script|style|code|pre';
+
+    _filter_url_escape_comments('', TRUE);
+    $text = is_null($text) ? '' : preg_replace_callback('`<!--(.*?)-->`s', '_filter_url_escape_comments', $text);
```
Comment (trackleft) [inline modules/custom/az_publication/src/Plugin/Filter/AZDOIFilter.php:32]: See Run PHP_CodeSniffer coding standards checks

https://app.probo.ci/#/dashboard/build/6e800054-6a00-4ca6-a9fc-777c3878bc1c/task/63368eb355c00007

**Comment by tadean** [comment]
@camikazegreen recommends including the DOI: prefix in the anchor label if one exists.

**Comment by tadean** [comment]
Added a check to verify that the filter format exists before applying it, as might occur on sites that have not yet applied distribution changes.

**Review by trackleft — DISMISSED** [review]

**Comment by trackleft** [comment]
This looks like the safest approach to me.

**Review by joeparsons — APPROVED** [review]

**Review by sanjanans — DISMISSED** [review]

**Review by joeparsons — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Review by camikazegreen — APPROVED** [review]
I reviewed this on a multidev of the pharmacy website and it works as expected.

**Review by sanjanans — APPROVED** [review]

**Pull request merged**